File: backend/src/main/java/com/example/Flicktionary/domain/movie/dto/MovieResponseWithDetail.java
package com.example.Flicktionary.domain.movie.dto;

import com.example.Flicktionary.domain.director.dto.DirectorDto;
import com.example.Flicktionary.domain.genre.dto.GenreDto;
import com.example.Flicktionary.domain.movie.entity.Movie;
import lombok.AllArgsConstructor;
import lombok.Getter;
import org.springframework.lang.NonNull;

import java.time.LocalDate;
import java.util.List;

@AllArgsConstructor
@Getter
public class MovieResponseWithDetail {
    @NonNull
    private final Long id;
    @NonNull
    private final Long tmdbId;
    @NonNull
    private final String title;
    private final String overview;
    private final LocalDate releaseDate;
    private final String posterPath;
    private final String status;
    private final Integer runtime;
    private final String productionCountry;
    private final String productionCompany;
    @NonNull
    private final double averageRating;
    @NonNull
    private final int ratingCount;
    private final List<GenreDto> genres;
    private final List<MovieCastDto> casts;
    private final DirectorDto director;

    public MovieResponseWithDetail(Movie movie) {
        this(movie.getId(),
                movie.getTmdbId(),
                movie.getTitle(),
                movie.getOverview(),
                movie.getReleaseDate(),
                movie.getPosterPath(),
                movie.getStatus(),
                movie.getRuntime(),
                movie.getProductionCountry(),
                movie.getProductionCompany(),
                movie.getAverageRating(),
                movie.getRatingCount(),
                movie.getGenres().stream().map(g -> new GenreDto(g.getId(), g.getName())).toList(),
                movie.getCasts().stream().map(MovieCastDto::new).toList(),
                movie.getDirector() == null ? null : new DirectorDto(movie.getDirector())
        );
    }
}

File: backend/src/main/java/com/example/Flicktionary/domain/movie/dto/MovieResponseWithDetail.kt
package com.example.Flicktionary.domain.movie.dto

import com.example.Flicktionary.domain.director.dto.DirectorDto
import com.example.Flicktionary.domain.genre.dto.GenreDto
import com.example.Flicktionary.domain.movie.entity.Movie
import java.time.LocalDate

data class MovieResponseWithDetail(
    val id: Long,
    val title: String,
    val overview: String,
    val releaseDate: LocalDate?,
    val posterPath: String?,
    val status: String,
    val runtime: Int,
    val productionCountry: String?,
    val productionCompany: String?,
    val averageRating: Double,
    val ratingCount: Int,
    val genres: List<GenreDto>,
    val casts: List<MovieCastDto>,
    val director: DirectorDto?
) {
    constructor(movie: Movie) : this(
        id = movie.id,
        title = movie.title,
        overview = movie.overview,
        releaseDate = movie.releaseDate,
        posterPath = movie.posterPath,
        status = movie.status,
        runtime = movie.runtime,
        productionCountry = movie.productionCountry,
        productionCompany = movie.productionCompany,
        averageRating = movie.averageRating,
        ratingCount = movie.ratingCount,
        genres = movie.genres.map { GenreDto(it.id, it.name) },
        casts = movie.casts.map { MovieCastDto(it) },
        director = movie.director?.let { DirectorDto(it) }
    )
}
